test/jdk/java/nio/channels/Channels/AsyncCloseStreams.java

```diff
@@ -0,0 +1,170 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
+ * @bug 8361495
+ * @summary Test for AsynchronousCloseException from uninterruptible FileChannel
+ * @run junit AsyncCloseStreams
+ */
+
+import java.io.Closeable;
+import java.io.InputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.channels.AsynchronousCloseException;
+import java.nio.channels.ClosedChannelException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.concurrent.LinkedTransferQueue;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class AsyncCloseStreams {
+    private static final Closeable STOP = () -> { };
+
+    private static Thread startCloseThread(LinkedTransferQueue<Closeable> q) {
+        return Thread.ofPlatform().start(() -> {
+            do {
+                try {
+                    Closeable c;
+                    if ((c = q.take()) != null) {
+                        if (c == STOP) {
+                            break;
+                        } else {
+                            try {
+                                c.close();
+                            } catch (IOException ignored) {
+                            }
+                        }
+                    }
+                } catch (InterruptedException ignored) {
+                    break;
+                }
+            } while (true);
+        });
+    }
+
+    @Test
+    public void available() throws InterruptedException, IOException {
+        var close = new LinkedTransferQueue<Closeable>();
+        Thread closeThread = startCloseThread(close);
+
+        try {
+            Path path = Files.createTempFile(Path.of("."), "foo", "bar");
+            path.toFile().deleteOnExit();
+
+            do {
+                InputStream in = Files.newInputStream(path);
+                close.offer(in);
+                int available = 0;
+                try {
+                    available = in.available();
+                } catch (AsynchronousCloseException ace) {
+                    System.out.println("AsynchronousCloseException caught");
+                    close.offer(STOP);
+                    break;
+                } catch (ClosedChannelException ignored) {
+                    continue;
+                } catch (Throwable t) {
+                    close.offer(STOP);
+                    fail("Unexpected error", t);
+                }
+                if (available < 0) {
+                    close.offer(STOP);
+                    fail("FAILED: available < 0");
+                }
+            } while (true);
+        } finally {
```

Comment:
If the finally blocks were to offer the STOP element then it would remove it from the 3 exit paths and make it more robust.